Grafana-CLI: Wrapper for `grafana-cli` within RPM/DEB packages and config/homepath are now global flags

gotjosh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our [`CONTRIBUTING.md`](https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md) guide.
2. Ensure you have added or ran the appropriate tests for your PR.
3. If it's a new feature or config option it will need a docs update. Docs are under the docs folder in repo root.
4. If the PR is unfinished, mark it as a draft PR.
5. Rebase your PR if it gets out of sync with master
6. Name your PR as `<FeatureArea>: Describe your change`. If it's a fix or feature relevant for changelog describe the user impact in the title. The PR title is used in the changelog for issues marked with `add to changelog` label. 
-->

**What this PR does / why we need it**:

When our users install the *nix packed version of grafana (via RPM or deb), tendency is to use the services and scripts installed as part of the package for grafana-server. These leverage the default configuration options by specifying several default paths.

This introduces a similar approach for the grafana-cli binary. We expose it through a wrapper to ensure a proper configuration is in place. 

To enable that, we tuck away into the `/usr/share/grafana` (default grafana home) directory the original binary (grafana-cli) and then use a bash script named grafana-cli as the wrapper.

In addition to that, this PR also:

- Makes the `--config` and `--homepath` flags global. Which allow us to significantly simplify the wrapper logic. 
- Adds a new  `--optionsString` which allow us to override default configuration options as a string. A thing to note is that this was and is still supported by passing these arguments as trailing args to the command, with the particularity that the trailing arguments take precedence over the flag to allow for backward compatibility. That means:

```bash
$ ./bin/darwin-amd64/grafana-cli -d --optionsString="cfg:default.paths.data=/dev/nullX cfg:default.paths.logs=/dev/nullX" --config=/dev/null admin reset-admin-password 1234 cfg:default.paths.plugins=/dev/nullY cfg:default.paths.logs=/dev/nullY
INFO[06-20|14:33:00] Config loaded from                       logger=settings file=/Users/josueabreu/go/src/github.com/grafana/grafana/conf/defaults.ini
INFO[06-20|14:33:00] Config loaded from                       logger=settings file=/dev/null
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.data=/dev/nullX"
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.logs=/dev/nullY"
INFO[06-20|14:33:00] Config overridden from command line      logger=settings arg="default.paths.plugins=/dev/nullY"
INFO[06-20|14:33:00] Path Home                                logger=settings path=/Users/josueabreu/go/src/github.com/grafana/grafana
INFO[06-20|14:33:00] Path Data                                logger=settings path=/dev/nullX
INFO[06-20|14:33:00] Path Logs                                logger=settings path=/dev/nullY
INFO[06-20|14:33:00] Path Plugins                             logger=settings path=/dev/nullY
INFO[06-20|14:33:00] Path Provisioning                        logger=settings path=/Users/josueabreu/go/src/github.com/grafana/grafana/conf/provisioning
INFO[06-20|14:33:00] App mode production                      logger=settings
INFO[06-20|14:33:00] Connecting to DB                         logger=sqlstore dbtype=sqlite3
INFO[06-20|14:33:00] Starting DB migration                    logger=migrator

Error: New password is too short

NAME:
   Grafana cli admin reset-admin-password - reset-admin-password <new password>

USAGE:
   Grafana cli admin reset-admin-password [arguments...
```
- Logs the paths configured as part of the command run to provide some feedback for the user if the `--debug` global flag is used

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

gotjosh committed:
Feature: Introduce a grafana-cli wrapper

When our users install the *nix packed version of grafana, tendency is to use the services and scripts installed as part of the package for grafana-server. These leverage the default configuration options by specifying the several default paths.

This introduces a similar approach for the grafana-cli binary. We exposed it through a wrapper to ensure a proper configuration is in place. To enable that, we add the .real suffix to the original binary (grafana-cli.real) and then use a bash script named grafana-cli as the wrapper.

gotjosh committed:
Make the config and homepath flags global

gotjosh committed:
Introduce optionsStrings as a global flag

This flag allows us to pass configuration overrides as a string.

The string follows the convention of configuration arguments separated by a space e.g. "cfg:default.paths.data=/dev/nullX cfg:default.paths.logs=/dev/nullX"

Also, it is backwards compatible with similar the previous configuration method through tailing arguments. Tailing arguments take presedence over the configuration options string.

gotjosh committed:
Only log configuration information in debug mode

gotjosh commented:
Superseds #17596

torkelo commented:
This looks like a nice improvement. 

We would need a bat/windows version as well.

gotjosh commented:
A bit of an update from me: Based on a conversation I had with @xlson, we were both under the impression that we currently have support for "muting" specific loggers based on this [configuration flag](https://github.com/grafana/grafana/blob/master/conf/sample.ini#L393). However, this is not the case - what this configuration does is supersed whatever level you originally specify on a particular logger.

gotjosh commented:
A follow up from my message above; I've added the option to mute logs as part of #17710

This PR now depends on that to silence the logs - the part that silences logs in debug only for the CLI is ready but will merge #17710 first before I push that code to this PR.

torkelo commented:
I can't see how this depends on the mute PR, how and why?

marefr commented:
> we were both under the impression that we currently have support for "muting" specific loggers based on this configuration flag. However, this is not the case - what this configuration does is supersed whatever level you originally specify on a particular logger.

Guess this is an existing problem with the cli (logs are being written of services starting up etc). I would recommend open a separate issue for resolving that if it's not critical for the success of this feature/fix.

torkelo commented:
Personally I think the cli should not connect to DB or use internal grafana-server config & services. It should talk to grafana-server via the HTTP api. 

Having the cli tool include big parts of the grafana-server runtime & config, and DB code feels very wrong to me. Wonder how other systems cli tools do this?

gotjosh commented:
Removed my comment re the dependency. Apologies for falling into a wrong assumption and deviating everyone. I'm with @marefr on this one, I'll create an issue for any other follow-ups.

gotjosh commented:
@torkelo I don't have any strong opinions either way. They both have trade-offs. 

I've seen both approaches work with great success in the wild although I have to admit that the latter (talk over HTTP) is more common these days in Go (due to the ability to test/integrate a bit easier).

xlson commented on build.go:
```diff
@@ -323,10 +328,16 @@ func createPackage(options linuxPackageOptions) {
+	// The grafana-cli binary is exposed through a wrapper to ensure a proper
+	// configuration is in place. To enable that, we need to store the original
+	// binary in a separate location to avoid conflicts.
+	runPrint("cp", "-p", filepath.Join(workingDir, "tmp/bin/"+cliBinary), filepath.Join(packageRoot, options.homeDir, cliBinary))
```
How about adding a `bin` folder to the home folder and putting it in there? Seems a bit cleaner to me.

gotjosh replied:
Sounds good to me.

xlson commented on pkg/cmd/grafana-cli/main.go:
```diff
@@ -51,6 +51,18 @@ func main() {
 			Name:  "debug, d",
 			Usage: "enable debug logging",
 		},
+		cli.StringFlag{
```
How about calling this `configOverrides` or something like that instead? Just `options` could work as well.

gotjosh replied:
Sounds good to me.

xlson reviewed (changes requested):
Looks great. Two minor things that me may want to change to. What do you think?

gotjosh committed:
Rename flag `optionsString` to `configOverrides`

xlson reviewed (approved):
LGTM!

gotjosh commented:
A head's up @xlson that I have removed the commit that moves the binary into a `/bin` folder. When I manually tested the change it doesn't seem to be creating the folder nor moving the binary which ends up erroring per the wrapper safeguard. 

We want to push this feature out asap as part of the new release and I believe the change to be non-critical. I have documented that change and two others we talked about as part of https://github.com/raintank/hosted-grafana/pull/96 to execute as a follow-up.

gotjosh committed:
Move the grafana-cli binary to $GRAFANA_HOME/bin

As part of the package install process, we copy all the release files and
directories into the grafana home directory. This includes the /bin folder
from where we copied the binaries into their respective destinations.
After that, the /bin folder gets deleted as we don't want to keep
duplicates of the binaries around.

As part of this commit, we moved the re-creation of /bin within
grafana-home and the copy of the original binary (again) after the
folder gets deleted.

gotjosh commented:
Managed to find and fix the problem @xlson. Merging this in.

gotjosh committed:
Update documentation to make reflect global flags change
